linkerd/stack/src/map_err.rs

```diff
@@ -1,58 +1,263 @@
+/// where `T` is a stack target.
+pub struct WrapFromTarget<T, E> {
+    target: T,
+    _err: PhantomData<fn(E)>,
+}
+
+// === impl NewMapErr===
+
+impl<W, X, N> NewMapErr<W, X, N> {
+    fn new(inner: N, extract: X) -> Self {
+        Self {
+            inner,
+            extract,
+            _cx: PhantomData,
+        }
+    }
+}
+
+impl<W, X: Clone, N> NewMapErr<W, X, N> {
+    /// A `NewService` layer that extracts a `WrapErr` implementation via the
+    /// `X`-typex `ExtractParam`.
+    pub fn layer_with(extract: X) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| Self::new(inner, extract.clone()))
+    }
+}
+
+impl<W, N> NewMapErr<W, (), N> {
+    /// A `NewService` layer that extracts a `W`-typed `WrapErr` implementation
+    /// from each target.
+    pub fn layer() -> impl layer::Layer<N, Service = Self> + Clone {
+        NewMapErr::layer_with(())
+    }
+}
+
+type ExtractWrapFromTarget<T, E> = fn(&T) -> WrapFromTarget<T, E>;
+
+/// A `NewService` layer that converts inner errors to an `E`-typed error via
+/// `From<(&T, Error)`, where `T` is a stack target.
+pub fn layer_new_from_target<E, T, N>(
+) -> impl layer::Layer<N, Service = NewMapErr<WrapFromTarget<T, E>, ExtractWrapFromTarget<T, E>, N>>
+       + Clone
+where
+    T: Clone,
+    WrapFromTarget<T, E>: WrapErr<Error> + Clone,
+{
+    NewMapErr::layer_with(
+        (|t: &T| WrapFromTarget {
+            target: t.clone(),
+            _err: PhantomData,
+        }) as ExtractWrapFromTarget<T, E>,
+    )
+}
+
+impl<T, W, X, N> NewService<T> for NewMapErr<W, X, N>
+where
+    X: ExtractParam<W, T>,
+    N: NewService<T>,
+{
+    type Service = MapErr<W, N::Service>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let wrap = self.extract.extract_param(&target);
+        let inner = self.inner.new_service(target);
+
+        MapErr { inner, wrap }
+    }
+}
+
+impl<W, X, N> Clone for NewMapErr<W, X, N>
+where
+    N: Clone,
+    X: Clone,
+{
+    fn clone(&self) -> Self {
+        Self::new(self.inner.clone(), self.extract.clone())
+    }
 }
 
-impl<S, F: Clone> MapErr<S, F> {
-    pub fn new(inner: S, f: F) -> Self {
-        Self { inner, f }
+impl<W, X, N> fmt::Debug for NewMapErr<W, X, N>
+where
+    N: fmt::Debug,
+    X: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self {
+            inner,
+            extract,
+            _cx,
+        } = self;
+        f.debug_struct("NewMapErr")
+            .field("inner", inner)
+            .field("extract", extract)
+            .finish()
+    }
+}
+
+// === impl MapErr ===
+
+impl<W, S> MapErr<W, S> {
+    pub fn new(inner: S, wrap: W) -> Self {
+        Self { inner, wrap }
+    }
+
+    pub fn layer(wrap: W) -> impl super::layer::Layer<S, Service = Self> + Clone
+    where
+        W: Clone,
+    {
+        super::layer::mk(move |inner| Self::new(inner, wrap.clone()))
```

Comment:
take it or leave it: it seems like `NewMapErr` could also have this constructor, using `CloneParam`? perhaps that should be the default `layer()` constructor to keep the "classic" `NewMapErr::layer(|error| { ... })` behavior, and the `ExtractParam` impls could be `layer_param` or something?

OTTOH we don't currently use a fn constructor with `NewMapErr`, so maybe it's not important...

Reply:
You can still use `MapErr::layer` for the original behavior afaiu.